buildSrc/src/main/groovy/org/elasticsearch/gradle/VersionCollection.groovy

```diff
@@ -20,164 +20,336 @@
+                safeAddToSet(foundVersion)
             }
         }
 
-        if (versions.empty) {
-            throw new GradleException("Unexpectedly found no version constants in Versions.java");
+        if (versionSet.empty) {
+            throw new GradleException("Unexpectedly found no version constants in Versions.java")
         }
 
-        /*
-         * The tip of each minor series (>= 5.6) is unreleased, so they must be built from source (we set branch to non-null), and we set
-         * the snapshot flag if and only if build.snapshot is true.
-         */
-        Version prevConsideredVersion = null
-        boolean found6xSnapshot = false
-        for (final int versionIndex = versions.size() - 1; versionIndex >= 0; versionIndex--) {
-            final Version currConsideredVersion = versions[versionIndex]
-
-            if (prevConsideredVersion == null
-                    || currConsideredVersion.major != prevConsideredVersion.major
-                    || currConsideredVersion.minor != prevConsideredVersion.minor) {
-
-                // This is a snapshot version. Work out its branch. NB this doesn't name the current branch correctly, but this doesn't
-                // matter as we don't BWC test against it.
-                String branch = "${currConsideredVersion.major}.${currConsideredVersion.minor}"
-
-                if (false == found6xSnapshot && currConsideredVersion.major == 6) {
-                    // TODO needs generalising to deal with when 7.x is cut, and when 6.x is deleted, and so on...
-                    branch = "6.x"
-                    found6xSnapshot = true
-                }
+        // If the major version has been released, then remove all of the alpha/beta/rc versions that exist in the set
+        versionSet.removeAll { it.suffix.isEmpty() == false && isMajorReleased(it, versionSet) }
 
-                versions[versionIndex] = new Version(
-                        currConsideredVersion.major, currConsideredVersion.minor,
-                        currConsideredVersion.revision, currConsideredVersion.suffix, buildSnapshot, branch)
-            }
+        // set currentVersion
+        Version lastVersion = versionSet.last()
+        currentVersion = new Version(lastVersion.major, lastVersion.minor, lastVersion.revision, lastVersion.suffix, buildSnapshot)
 
-            if (currConsideredVersion.onOrBefore("5.6.0")) {
-                break
-            }
+        // remove all of the potential alpha/beta/rc from the currentVersion
+        versionSet.removeAll {
+            it.suffix.isEmpty() == false &&
+            it.major == currentVersion.major &&
+            it.minor == currentVersion.minor &&
+            it.revision == currentVersion.revision }
 
-            prevConsideredVersion = currConsideredVersion
+        // re-add the currentVersion to the set
+        versionSet.add(currentVersion)
```

Comment:
Why would current be in bwc versions at all?

Reply:
discussed w/ you in person about the test failures, would love to fix up in another PR